libraries/chain/hotstuff/finalizer.cpp

```diff
@@ -0,0 +1,302 @@
+      fsi.last_vote_range_start = p_start;
+
+      if (chain.b1 && chain.b1->timestamp() > fsi.lock.timestamp)
+         fsi.lock = proposal_ref(chain.b1);                // commit phase on b1
+
+      decision = enough_for_strong_vote ? VoteDecision::StrongVote : VoteDecision::WeakVote;
+   } else {
+      dlog("bsp_last_qc=${bsp}, last_qc_block_num=${lqc}, fork_db root block_num=${f}",
+           ("bsp", !!bsp_last_qc)("lqc",!!p->last_qc_block_num())("f",fork_db.root()->block_num()));
+      if (p->last_qc_block_num())
+         dlog("last_qc_block_num=${lqc}", ("lqc", *p->last_qc_block_num()));
+   }
+   if (decision != VoteDecision::NoVote)
+      dlog("Voting ${s}", ("s", decision == VoteDecision::StrongVote ? "strong" : "weak"));
+   return decision;
+}
+
+// ----------------------------------------------------------------------------------------
+std::optional<vote_message> finalizer::maybe_vote(const bls_public_key& pub_key, const block_state_ptr& p,
+                                                  const digest_type& digest, const fork_database_if_t& fork_db) {
+   finalizer::VoteDecision decision = decide_vote(p, fork_db);
+   if (decision == VoteDecision::StrongVote || decision == VoteDecision::WeakVote) {
+      bls_signature sig;
+      if (decision == VoteDecision::WeakVote) {
+         // if voting weak, the digest to sign should be a hash of the concatenation of the finalizer_digest
+         // and the string "WEAK"
+         sig =  priv_key.sign(create_weak_digest(digest));
+      } else {
+         sig =  priv_key.sign({(uint8_t*)digest.data(), (uint8_t*)digest.data() + digest.data_size()});
+      }
+      return vote_message{ p->id(), decision == VoteDecision::StrongVote, pub_key, sig };
+   }
+   return {};
+}
+
+// ----------------------------------------------------------------------------------------
+void finalizer_set::save_finalizer_safety_info() const {
+
+   if (!persist_file.is_open()) {
+      EOS_ASSERT(!persist_file_path.empty(), finalizer_safety_exception,
+                 "path for storing finalizer safety information file not specified");
+      if (!std::filesystem::exists(persist_file_path.parent_path()))
+          std::filesystem::create_directories(persist_file_path.parent_path());
+      persist_file.set_file_path(persist_file_path);
+      persist_file.open(fc::cfile::truncate_rw_mode);
+      EOS_ASSERT(persist_file.is_open(), finalizer_safety_exception,
+                 "unable to open finalizer safety persistence file: ${p}", ("p", persist_file_path));
+   }
+   try {
+      persist_file.seek(0);
+      fc::raw::pack(persist_file, fsi_t::magic);
+      fc::raw::pack(persist_file, (uint64_t)(finalizers.size() + inactive_safety_info.size()));
+      for (const auto& [pub_key, f] : finalizers) {
+         fc::raw::pack(persist_file, pub_key);
+         fc::raw::pack(persist_file, f.fsi);
+      }
+      if (!inactive_safety_info_written) {
```

Comment:
Can we have a separate file for inactive safety info so it won't be written every time?

Reply:
It is not written every time.